libsweep/src/sweep.cc

```diff
@@ -6,29 +6,38 @@
 #include "sweep.h"
 
 #include <chrono>
+#include <exception>
 #include <string>
 #include <thread>
 
 int32_t sweep_get_version(void) { return SWEEP_VERSION; }
 bool sweep_is_abi_compatible(void) { return sweep_get_version() >> 16u == SWEEP_VERSION_MAJOR; }
 
-typedef struct sweep_error { std::string what; } sweep_error;
-
-typedef struct sweep_device {
-  sweep::serial::device_s serial; // serial port communication
-  bool is_scanning;
-  sweep::queue::queue<sweep_scan_s> scan_queue;
-  std::atomic<bool> stop_thread;
-} sweep_device;
+struct sweep_error {
+  std::string what;
+};
 
 #define SWEEP_MAX_SAMPLES 4096
 
-typedef struct sweep_scan {
+struct sweep_scan {
   int32_t angle[SWEEP_MAX_SAMPLES];           // in millidegrees
   int32_t distance[SWEEP_MAX_SAMPLES];        // in cm
   int32_t signal_strength[SWEEP_MAX_SAMPLES]; // range 0:255
   int32_t count;
-} sweep_scan;
+};
+
+struct sweep_device {
+  sweep::serial::device_s serial; // serial port communication
+  bool is_scanning;
+  std::atomic<bool> stop_thread;
+
+  struct Element {
+    std::unique_ptr<sweep_scan> scan;
+    std::exception_ptr error;
+  };
```

Comment:
currentl, a `sweep_scan` object is a struct of size ~12KB irrespective of how many actual scans it contains, so you certainly don't want it be copied around more than absolutely necessary. I actually meant to replace the current implementation of `sweep_scan` by a vector (essentaillly the same as the representation in the public c+ api), but I didn't want to mix up those two changes in a single PR.